File: Week27 Codes/6.Signing/Signature.py
import botan3 as botan
from base64 import b64encode

def main():
    rng = botan.RandomNumberGenerator()

    # Generate RSA private key
    rsa_priv_key = botan.PrivateKey.create("RSA", "2048", rng)

    # Extract public key from private key
    rsa_pub_key = rsa_priv_key.get_public_key()

    # Message to sign
    message = "Hello, this is a signed message."

    # Sign the message
    signer = botan.PKSign(rsa_priv_key, 'EMSA-PKCS1-v1_5(SHA-256)')
    signer.update(message)
    signature = signer.finish(rng)

    # Hex encode the signature
    hex_signature = b64encode(signature).decode()

    # Verify the signature
    verifier = botan.PKVerify(rsa_pub_key, 'EMSA-PKCS1-v1_5(SHA-256)')
    verifier.update(message)
    valid = verifier.check_signature(signature)

    # Output the results
    print("Message:", message)
    print("Signature:", hex_signature)
    print("Verification:", "valid" if valid else "invalid")

if __name__ == "__main__":
    main()

File: code/Week27 Codes/6.Signing/Signature.cpp
#include <botan/auto_rng.h>
#include <botan/hash.h>
#include <botan/pkcs8.h>
#include <botan/pubkey.h>
#include <botan/rsa.h>
#include <botan/x509_key.h>
#include <botan/hex.h>
#include <iostream>

int main() {
    Botan::AutoSeeded_RNG rng;

    // Generate RSA private key
    Botan::RSA_PrivateKey private_key(rng, 2048);

    // Extract public key from private key
    Botan::RSA_PublicKey public_key(private_key);

    // Message to sign
    std::string message = "Hello, this is a signed message.";

    // Hash the message
    std::unique_ptr<Botan::HashFunction> hash_function = Botan::HashFunction::create("SHA-256");
    hash_function->update(reinterpret_cast<const uint8_t*>(message.data()), message.size());
    std::vector<uint8_t> message_hash = hash_function->final_stdvec();

    // Sign the hash
    Botan::PK_Signer signer(private_key, rng, "EMSA-PKCS1-v1_5(SHA-256)");
    std::vector<uint8_t> signature = signer.sign_message(message_hash, rng);

    // Hex encode the signature
    std::string hex_signature = Botan::hex_encode(signature);

    // Verify the signature
    Botan::PK_Verifier verifier(public_key, "EMSA-PKCS1-v1_5(SHA-256)");
    bool valid = verifier.verify_message(message_hash, signature);

    // Output the results
    std::cout << "Message: " << message << std::endl;
    std::cout << "Signature: " << hex_signature << std::endl;
    std::cout << "Verification: " << (valid ? "valid" : "invalid") << std::endl;

    return 0;
}